Initial Version of Couchbase driver

babltiga opened the pull request:
## Issue reference

New Driver Implementation

## Documentation PR reference

https://github.com/mongock/mongock-docs/pull/42

## Examples PR reference

https://github.com/mongock/mongock-examples/pull/50

## Description and context

Couchbase Driver implementation supporting Couchbase Server 6+ (may also support the earlier once but tests are executed against v6 and v7).

## Benefits

New Driver backed by Couchbase server.

## Possible Drawbacks

None

## Checklist 
- [x] Unit tests provided
- [x] Integration tests provided 
- [x] Documentation updated in [documentation project](https://github.com/mongock/mongock-docs)
- [x] Updated example in [example projects](https://github.com/mongock/mongock-examples)

babltiga committed:
Initial Version of Couchbase driver

osantana85 commented on drivers/couchbase/README.md:
```diff
@@ -0,0 +1 @@
+# Parent project for all the community Couchbase drivers
```
This is an additional comment, because it's needed to add the following dependency to the  module "mongock-community-bom" (pom.xml):

```
      <!-- Couchbase community -->
      <dependency>
        <groupId>io.mongock</groupId>
        <artifactId>mongock-driver-couchbase-bom</artifactId>
        <version>${project.version}</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
```

babltiga replied:
@osantana85 can you elaborate what should be done here ?
Just add more description into the README.md file? Or I can just share the link to the official docs?

osantana85 replied:
Hello @babltiga, sorry for the confusion. This comment is here because you need to add the dependency to a file that was not modified by this PR (mongock-community-bom > pom.xml). You don't need to modify the README.md file.

osantana85 commented on drivers/couchbase/mongock-driver-couchbase-bom/pom.xml:
```diff
@@ -0,0 +1,40 @@
+      </dependency>
+      <dependency>
+        <groupId>io.mongock</groupId>
+        <artifactId>couchbase-springdata-driver</artifactId>
```
Should be: `<artifactId>couchbase-springboot-driver</artifactId>`

osantana85 commented on drivers/couchbase/couchbase-driver/src/main/java/io/mongock/driver/couchbase/repository/CouchbaseLockRepository.java:
```diff
@@ -0,0 +1,97 @@
+      LockEntry existingLock = new CouchbaseLockEntry(result.contentAsObject());
+      if(newLock.getOwner().equals(existingLock.getOwner()) ||
+          new Date().after(existingLock.getExpiresAt())){
+        logger.debug("Lock with key {} already owned by us or is expired, so trying to preform a lock.", existingLock.getKey());
```
Typo: preform > perform

osantana85 commented on drivers/couchbase/couchbase-driver/src/main/java/io/mongock/driver/couchbase/repository/CouchbaseLockRepository.java:
```diff
@@ -0,0 +1,97 @@
+        throw new LockPersistenceException("Get By" + newLock.getKey(), newLock.toString(), "Still locked by " + existingLock.getOwner() + " until " + existingLock.getExpiresAt());
+      }
+    } catch (DocumentNotFoundException documentNotFoundException){
+      logger.debug("Lock with key {} does not exist, so trying to preform a lock.", newLock.getKey());
```
Typo: preform > perform

osantana85 commented on drivers/couchbase/couchbase-driver/src/main/java/io/mongock/driver/couchbase/repository/CouchbaseLockRepository.java:
```diff
@@ -0,0 +1,97 @@
+      GetResult result = collection.get(newLock.getKey());
+      LockEntry existingLock = new CouchbaseLockEntry(result.contentAsObject());
+      if(newLock.getOwner().equals(existingLock.getOwner())){
+        logger.debug("Lock with key {} already owned by us, so trying to preform a lock.", existingLock.getKey());
```
Typo: preform > perform

osantana85 commented on drivers/couchbase/couchbase-driver/src/main/java/io/mongock/driver/couchbase/repository/CouchbaseRepositoryBase.java:
```diff
@@ -0,0 +1,195 @@
+      try{
+        document.put(key, value);  
+      } catch (IllegalArgumentException e){
+        throw new MongockException("Unsupported Couchbase type " + value.getClass().getName());
```
Possible NullPointerException (value could be null). You should consider check null value before the first if statement.

babltiga replied:
@osantana85 if the value is null, it should never reach line 186, as IllegalArgumentException won't be thrown, null is a valid value for the Couchbase.

osantana85 replied:
Ok, there was a warning in the IDE about it.

osantana85 commented on drivers/couchbase/couchbase-springboot-driver/src/main/java/io/mongock/driver/couchbase/springboot/config/CouchbaseSpringbootContext.java:
```diff
@@ -0,0 +1,51 @@
+  }
+  
+  private boolean isCustomCollection(CouchbaseConfiguration couchbaseConfiguration){
+    return StringUtils.hasText(couchbaseConfiguration.getCollection())  &&
```
Could we check somehow if server allows scopes/collections (v7+) here? That way we could validate the provided configuration. If not, the server will fail itself.

babltiga replied:
@osantana85 I do not think it is a good idea to verify anything based on the Couchbase version alone, as technically the SDK client can be on couchbase 6.6+ enterprise with developer preview mode enabled. https://docs.couchbase.com/server/6.6/developer-preview/collections/collections-overview.html 

But if you think it makes sense, I can easily throw an BeanInstantiationException in case there is custom scope or collection and the version is < 7.

osantana85 replied:
Ok, in that case it will be better to don't validate it. I known you'd have a good reason ;)

osantana85 commented on drivers/couchbase/couchbase-driver/src/main/java/io/mongock/driver/couchbase/repository/ChangeEntryKeyGenerator.java:
```diff
@@ -0,0 +1,17 @@
+
+public class ChangeEntryKeyGenerator {
+
+  public static final String KEY_PREFIX = "mongock-";
```
We suggest to use "mongockChangeEntry-" prefix for ChangeEntry and to use "mongockLock-" prefix for LockEntry. (It's needed to add a LockEntryKeyGenerator for LockEntry).

osantana85 replied:
ChangeEntry is ok, but LockEntry hasn't prefix. Currently Lock entries are created with "DEFAULT_KEY" value (it's Mongock's default value) as Couchbase document key, and we think the Couchbase document key should be something like "mongockLockEntry-DEFAULT_KEY", to prevent conflicts with users documents.

osantana85 commented on drivers/couchbase/couchbase-driver/src/main/java/io/mongock/driver/couchbase/repository/CouchbaseRepositoryBase.java:
```diff
@@ -0,0 +1,195 @@
+  private final static Logger logger = LoggerFactory.getLogger(CouchbaseRepositoryBase.class);
+  private final static String INDEX_NAME = "idx_mongock_keys";
+  protected final static String DOCUMENT_TYPE = "mongock";
+  protected final static String DOCUMENT_TYPE_KEY = "_type";
```
ChangeEntry already include a "type" field. Should we consider to use a different field name for this one?

osantana85 commented on drivers/couchbase/couchbase-driver/src/main/java/io/mongock/driver/couchbase/repository/CouchbaseRepositoryBase.java:
```diff
@@ -0,0 +1,195 @@
+public class CouchbaseRepositoryBase <DOMAIN_CLASS> implements EntityRepository<DOMAIN_CLASS, JsonObject>, RepositoryIndexable {
+  private final static Logger logger = LoggerFactory.getLogger(CouchbaseRepositoryBase.class);
+  private final static String INDEX_NAME = "idx_mongock_keys";
+  protected final static String DOCUMENT_TYPE = "mongock";
```
This document type is currently used only for ChangeEntry. We suggest to rename it to "mongockChangeEntry" and add another one named "mongockLockEntry" for LockEntry documents (currenty LockEntry documents don't include "_type" field).

osantana85 reviewed (changes requested):
Hello @babltiga, great job! I added a few comments, most of them minor changes. If you need more details please let me know. Thanks!

babltiga committed:
Fixed typos in the logs.
Changed the couchbase key prefix.
Change the doctype file name.
Changed the doctype key value for change sets.

osantana85 commented on drivers/couchbase/couchbase-driver/src/main/java/io/mongock/driver/couchbase/repository/CouchbaseRepositoryBase.java:
```diff
@@ -35,8 +35,8 @@
   private final static String INDEX_NAME = "idx_mongock_keys";
-  protected final static String DOCUMENT_TYPE = "mongock";
-  protected final static String DOCUMENT_TYPE_KEY = "_type";
+  protected final static String DOCUMENT_TYPE_CHANGE_ENTRY = "mongockChangeEntry";
```
This ChangeEntry specific code should be moved to `CouchbaseChangeEntryRepository`.
Additionally, for LockEntry entity we should add another document type, named "mongockLockEntry", and assign it in its repository `CouchbaseLockRepository`.

osantana85 commented on drivers/couchbase/couchbase-driver/src/main/java/io/mongock/driver/couchbase/repository/CouchbaseRepositoryBase.java:
```diff
@@ -51,7 +51,7 @@ public CouchbaseRepositoryBase(Cluster cluster, Collection collection, Set<Strin
   public JsonObject mapFieldInstances(List<FieldInstance> fieldInstanceList) {
     JsonObject document = JsonObject.create();
-    document.put(DOCUMENT_TYPE_KEY, DOCUMENT_TYPE);
+    document.put(DOCUMENT_TYPE_KEY, DOCUMENT_TYPE_CHANGE_ENTRY);
```
This ChangeEntry specific code should be moved to `CouchbaseChangeEntryRepository`.
Additionally, for LockEntry entity we should add another document type, named "mongockLockEntry", and assign it in its repository `CouchbaseLockRepository`.

osantana85 commented on drivers/couchbase/couchbase-driver/src/main/java/io/mongock/driver/couchbase/util/CouchbaseVersionUtil.java:
```diff
@@ -0,0 +1,55 @@
+import com.couchbase.client.java.Bucket;
+import com.couchbase.client.java.manager.bucket.GetBucketOptions;
+
+public final class CouchbaseVersionUtil {
```
I think this file is useless if we don't validate server in the startup, but if would be useful in if the near future we'll decide to check it. So it's fine.

babltiga replied:
Yep, let's keep it for now, in case we do not need it we can just get rid of it whenever we want. There should be almost 0 effort to maintain this "util".

osantana85 reviewed (changes requested):
Hello @babltiga , thanks for your changes. Most of the requests are solved. I've added a few new comments. Good job!

babltiga committed:
Set the _docType for both the change and lock entries.
Add missing dependency into community-bom.

osantana85 reviewed (changes requested):
Hello @babltiga , sorry for the delay, but this week we have been a bit busy releasing Mongock's version 5.2.4 (solving a few issues).
The PR is almost ok to merge. I added a comment related to LockEntry documents key prefix.
Additionally, could you rebase/merge latest Mongock's changes to your PR branch? Currently version of modules should be 5.2.5-SNAPSHOT (now yours are 5.2.3-SNAPSHOT).
Thanks in advance.

babltiga committed:
Merge branch 'develop' into couchbase-driver

babltiga committed:
prefix lock entry keys with 'mongockLockEntry'

osantana85 reviewed (approved):
Hello @babltiga , your PR is approved to merge. Thanks for your great job.
